OrbitLinuxTracing/ThreadStateVisitor.cpp

```diff
@@ -0,0 +1,276 @@
+    ERROR("Processed sched:sched_wakeup for thread %d but unexpected previous state %s", tid,
+          ThreadStateSlice::ThreadState_Name(open_state.state));
+  }
+
+  ThreadStateSlice slice;
+  slice.set_tid(tid);
+  slice.set_thread_state(open_state.state);
+  slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+  slice.set_end_timestamp_ns(timestamp_ns);
+  tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+  return slice;
+}
+
+std::optional<ThreadStateSlice> ThreadStateManager::OnSchedSwitchIn(uint64_t timestamp_ns,
+                                                                    pid_t tid) {
+  static constexpr ThreadStateSlice::ThreadState kNewState = ThreadStateSlice::kRunning;
+
+  auto open_state_it = tid_open_states_.find(tid);
+  if (open_state_it == tid_open_states_.end()) {
+    ERROR("Processed sched:sched_switch(in) but previous state of thread %d is unknown", tid);
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  const OpenState& open_state = open_state_it->second;
+  if (timestamp_ns < open_state.begin_timestamp_ns) {
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  if (open_state.state == kNewState) {
+    // No state change: do nothing and don't overwrite the previous begin timestamp.
+    return std::nullopt;
+  }
+
+  // Don't print an error even if open_state.state != ThreadStateSlice::kRunnable: it seems to be
+  // sometimes possible for a thread to go from a non-runnable state directly to running, skipping
+  // the sched:sched_wakeup event.
+
+  ThreadStateSlice slice;
+  slice.set_tid(tid);
+  slice.set_thread_state(open_state.state);
+  slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+  slice.set_end_timestamp_ns(timestamp_ns);
+  tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+  return slice;
+}
+
+std::optional<ThreadStateSlice> ThreadStateManager::OnSchedSwitchOut(
+    uint64_t timestamp_ns, pid_t tid, ThreadStateSlice::ThreadState new_state) {
+  auto open_state_it = tid_open_states_.find(tid);
+  if (open_state_it == tid_open_states_.end()) {
+    ERROR("Processed sched:sched_switch(out) but previous state of thread %d is unknown", tid);
+    tid_open_states_.insert_or_assign(tid, OpenState{new_state, timestamp_ns});
+    return std::nullopt;
+  }
+
+  const OpenState& open_state = open_state_it->second;
+  if (timestamp_ns < open_state.begin_timestamp_ns) {
+    tid_open_states_.insert_or_assign(tid, OpenState{new_state, timestamp_ns});
+    return std::nullopt;
+  }
+
+  // As we are switching out of a CPU, if the previous state was kRunnable, assume it was kRunning.
+  // This is because when we retrieve the initial thread states we have no way to distinguish
+  // between kRunnable and kRunning. After all, "running" is not a real thread state.
```

Comment:
Same question here: Is "running" not a real thread state, or is "runnable" not a real thread state? (TASK_RUNNING seems to be a real state the Linux scheduler code.)